Failure to install go in e2e workflows

### Which jobs are flaking?

I think this affects all e2e jobs since they all install go the same way.
Here is one example of a failing job: https://github.com/metal3-io/baremetal-operator/actions/runs/21616532469

### Which tests are flaking?

Installation of go that happens before the tests actually start.

### Since when has it been flaking?

First occurrence I have seen was on 1st February: https://github.com/metal3-io/baremetal-operator/actions/runs/21557005916/job/62115095091

### Jenkins link

_No response_

### Reason for failure (if possible)

There is some kind of issue downloading and extracting the archive. Probably it was corrupted during download.

```
+ /home/ubuntu/_work/baremetal-operator/baremetal-operator/hack/e2e/ensure_go.sh
+ MINIMUM_GO_VERSION=go1.24.12
+ verify_go_version
++ command -v go
+ '[' -x '' ']'
+ [[ linux-gnu == \l\i\n\u\x\-\g\n\u ]]
+ echo 'go not found, installing'
+ curl -sLo /tmp/go1.24.12.linux-amd64.tar.gz https://go.dev/dl/go1.24.12.linux-amd64.tar.gz
go not found, installing
+ sudo tar -C /usr/local -xzf /tmp/go1.24.12.linux-amd64.tar.gz

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
Error: Process completed with exit code 2.
```

### Anything else we need to know?

As a fix/workaround we could add a retry mechanism and file integrity check.

### Label(s) to be applied

/kind flake
One or more /area label. See https://github.com/metal3-io/baremetal-operator/labels for the list of labels.

--- a/hack/e2e/ensure_go.sh
+++ b/hack/e2e/ensure_go.sh
@@ -18,13 +18,20 @@ verify_go_version() {
         fi
         if [[ "${OSTYPE}" == "linux-gnu" ]]; then
             echo "go not found, installing"
+            local GO_TARBALL="${MINIMUM_GO_VERSION}.linux-amd64.tar.gz"
+            for attempt in 1 2 3; do
+                echo "Downloading Go (attempt ${attempt}/3)"
+                if curl -fsSL -o "/tmp/${GO_TARBALL}" "https://go.dev/dl/${GO_TARBALL}" && \
+                    curl -fsSL -o "/tmp/${GO_TARBALL}.sha256" "https://dl.google.com/go/${GO_TARBALL}.sha256" && \
+                    echo "$(cat "/tmp/${GO_TARBALL}.sha256")  /tmp/${GO_TARBALL}" | sha256sum --check --quiet; then
+                    break
+                fi
+                rm -f "/tmp/${GO_TARBALL}" "/tmp/${GO_TARBALL}.sha256"
+            done
+            [[ -f "/tmp/${GO_TARBALL}" ]] || { echo "ERROR: failed to download valid Go archive"; return 2; }
             set -x
-            curl -sL \
-                -o "/tmp/${MINIMUM_GO_VERSION}.linux-amd64.tar.gz" \
-                "https://go.dev/dl/${MINIMUM_GO_VERSION}.linux-amd64.tar.gz"
-            sudo tar \
-                -C /usr/local \
-                -xzf "/tmp/${MINIMUM_GO_VERSION}.linux-amd64.tar.gz"
+            sudo tar -C /usr/local -xzf "/tmp/${GO_TARBALL}"
+            rm -f "/tmp/${GO_TARBALL}" "/tmp/${GO_TARBALL}.sha256"
             set +x
             export PATH="${PATH}:/usr/local/go/bin"
             GO="$(command -v go)"

